concordium-std/src/lib.rs

```diff
@@ -242,7 +250,61 @@
 //! Other error codes may be added in the future and custom error codes should
 //! not use the range `i32::MIN` to `i32::MIN + 100`.
 //!
+//! # Collections
+//!
+//! Several collections are available for use in a smart contract and choosing
+//! the right one can result in significant cost savings.
+//!
+//! First off, Rust's own standard library provides [several efficient data structures](https://doc.rust-lang.org/std/collections/)
+//! which can be used in a smart contract.
+//!
+//! However, these can become costly for collections holding a large number of
+//! elements, which needs to be persisted in the smart contract state across
+//! contract updates.
+//! The reason being these data structures are designed to be kept in-memory and
+//! persisting them means reading and writing the entire collection to a single
+//! entry in the smart contract key-value store on every contract update.
+//! This is wasteful when only part of the collection is relevant on each
+//! update.
+//!
+//! In the above mentioned scenarios, it is instead recommended to use one of
+//! the smart contract tailored collections, as these partisions the collection
+//! into multiple key-value stores, resulting in cheaper read and writes to the
+//! state.
+//!
+//! The collections can be grouped as:
+//! - Maps: [`StateMap`], [`StateBTreeMap`]
+//! - Sets: [`StateSet`], [`StateBTreeSet`]
+//!
+//! ## When should you use which collection?
+//!
+//! This section presents a rough guideline for when to reach for each of the
+//! collections.
+//!
+//! ### Use `StateMap` when:
+//!
+//! - You want to track which keys you have seen.
+//! - Arbitrary values are associated with each of the keys.
+//!
+//! ### Use `StateBTreeMap` when:
+//!
+//! - You want to track which keys you have seen.
+//! - Arbitrary values are associated with each of the keys.
+//! - The keys have some _ordering_ which is relevant.
```

Comment:
Maybe give an example here, "for example when you wish to query the smallest key".